### PR TITLE
Preserve state on SDK switch (#22430)

### LINKED ISSUE #22430: [Task]: Playground state preservation on SDK switch
### What needs to happen?

Currently Playground maintains a single state of a document. When SDK changes, the code in the editor is lost, and the default example for the given SDK is loaded.

Need to maintain a separate state for each SDK so that:

1. The Playground is started with language A.
2. The Playground loads the snippet for A.
3. The user modifies the snippet code.
4. The user switches the language to B.
5. The Playground loads the snippet for B.
6. The user switches the language back to A.
7. The in-memory snippet for A preserves the earlier modification by the user.

### Issue Priority

Priority: 3

### Issue Component

Component: beam-playground

### PR DESCRIPTION
**Pull request by alexeyinkin**
Resolves #22430

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by alexeyinkin** [comment]
R: @damondouglas

File: playground/frontend/lib/pages/playground/states/playground_state.dart [hunk 1]
```diff
@@ -45,95 +49,142 @@ const kCachedResultsLog =
...
+        loadDefaultIfNot: loadDefaultIfNot);
 
     if (notify) {
       notifyListeners();
     }
   }
 
-  set source(String source) {
-    codeController.text = source;
+  void setSource(String source) {
```
Comment (alexeyinkin) [inline playground/frontend/lib/pages/playground/states/playground_state.dart:185]: We recently turned this from `setSource` to `set source` following the linter, but now we need nullable getter with a non-nullable setter, and that does not work, so back to the method again.

File: playground/frontend/lib/pages/playground/states/examples_state.dart [hunk 2]
```diff
@@ -173,31 +173,27 @@ class ExampleState with ChangeNotifier {
...
-            // First parameter is an empty string, because we don't need path to get the default example.
-            GetExampleRequestWrapper('', value),
-          ),
-        ),
-      );
+    try {
+      await Future.wait(SDK.values.map(_loadDefaultExample));
+    } catch (ex) {
+      if (defaultExamplesMap.isEmpty) {
+        rethrow;
```
Comment (alexeyinkin) [inline playground/frontend/lib/pages/playground/states/examples_state.dart:180]: We used to load the 4 default examples sequentially, and if any of them failed, nothing was loaded. Now go in parallel and allow failures in some of them.

File: playground/frontend/lib/pages/embedded_playground/components/embedded_editor.dart [hunk 3]
```diff
@@ -29,11 +31,17 @@ class EmbeddedEditor extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
     final state = Provider.of<PlaygroundState>(context);
+    final controller = state.snippetEditingController;
+
+    if (controller == null) {
+      return const LoadingIndicator(size: kLgLoadingIndicatorSize);
```
Comment (alexeyinkin) [inline playground/frontend/lib/pages/embedded_playground/components/embedded_editor.dart:37]: We used to show a blank editor before the example is loaded, and it was confusing. Now we only get the controller with the example already in it, unless deliberately showing an empty editor.

**Review by damondouglas — APPROVED** [review]
@pabloem LGTM

**Comment by pabloem** [comment]
LGTM thanks

**Pull request merged**